src/media_tools/av_parsers.c

```diff
@@ -1545,6 +1545,595 @@ GF_Err gf_m4a_write_config(GF_M4ADecSpecInfo *cfg, char **dsi, u32 *dsi_size)
+		*frame_size = 0;
+		return GF_NON_COMPLIANT_BITSTREAM;
+	}
+
+	gf_bs_read_u64(bs); //skip
+
+	return GF_OK;
+}
+
+static GF_Err av1_parse_obu_header(GF_BitStream *bs, ObuType *obu_type, Bool *obu_extension_flag, Bool *obu_has_size_field, u8 *temporal_id, u8 *spatial_id)
+{
+	Bool forbidden = gf_bs_read_int(bs, 1);
+	if (forbidden) {
+		return GF_NON_COMPLIANT_BITSTREAM;
+	}
+	assert(obu_type && obu_extension_flag && obu_has_size_field && temporal_id && spatial_id);
+	*obu_type = gf_bs_read_int(bs, 4);
+	*obu_extension_flag = gf_bs_read_int(bs, 1);
+	*obu_has_size_field = gf_bs_read_int(bs, 1);
+	gf_bs_read_int(bs, 1); /*obu_reserved_1bit*/
+	if (*obu_extension_flag) {
+		*temporal_id = gf_bs_read_int(bs, 3);
+		*spatial_id = gf_bs_read_int(bs, 2);
+		/*extension_header_reserved_3bits = */gf_bs_read_int(bs, 3);
+	}
+
+	return GF_OK;
+}
+
+const char *av1_get_obu_name(ObuType obu_type)
+{
+	switch (obu_type) {
+	case OBU_SKIP: return "skip";
+	case OBU_SEQUENCE_HEADER: return "seq_header";
+	case OBU_TEMPORAL_DELIMITER: return "delimiter";
+	case OBU_FRAME_HEADER: return "frame_header";
+	case OBU_TILE_GROUP: return "tile_group";
+	case OBU_METADATA: return "metadata";
+	case OBU_FRAME: return "frame";
+	case OBU_REDUNDANT_FRAME_HEADER: return "redundant_frame_header";
+	case OBU_PADDING: return "padding";
+	default: return "unknown";
+	}
+}
+
+Bool av1_is_obu_header(ObuType obu_type) {
+	switch (obu_type) {
+	case OBU_SEQUENCE_HEADER:
+	case OBU_METADATA:
+		return GF_TRUE;
+	default:
+		return GF_FALSE;
+	}
+}
+
+static Bool av1_is_obu_frame(ObuType obu_type) {
+	switch (obu_type) {
+	case OBU_TEMPORAL_DELIMITER:
+	case OBU_PADDING:
+	case OBU_REDUNDANT_FRAME_HEADER:
+		return GF_FALSE;
+	default:
+		return GF_TRUE;
+	}
+}
+
+u64 leb128(GF_BitStream *bs, u8 *opt_Leb128Bytes) {
+	u64 value = 0;
+	u8 Leb128Bytes = 0, i = 0;
+	for (i = 0; i < 8; i++) {
+		u8 leb128_byte = gf_bs_read_u8(bs);
+		value |= ((leb128_byte & 0x7f) << (i * 7));
+		Leb128Bytes += 1;
+		if (!(leb128_byte & 0x80)) {
+			break;
+		}
+	}
+
+	if (opt_Leb128Bytes) {
+		*opt_Leb128Bytes = Leb128Bytes;
+	}
+	return value;
+}
+
+#define OBU_COPY \
+	gf_bs_seek(bs, pos); \
+	GF_AV1_OBUArrayEntry *a; \
+	GF_SAFEALLOC(a, GF_AV1_OBUArrayEntry); \
+	a->obu = gf_malloc((size_t)obu_length); \
+	gf_bs_read_data(bs, a->obu, (u32)obu_length); \
+	a->obu_length = obu_length; \
+	a->obu_type = obu_type;
+
+static void av1_populate_state_from_obu(GF_BitStream *bs, u64 pos, u64 obu_length, ObuType obu_type, AV1State *state)
+{
+	if (av1_is_obu_header(obu_type)) {
+		OBU_COPY;
+		if (!state->frame_state.header_obus) state->frame_state.header_obus = gf_list_new();
+		gf_list_add(state->frame_state.header_obus, a);
+	}
+	if (av1_is_obu_frame(obu_type)) {
+		OBU_COPY;
+		if (!state->frame_state.frame_obus) state->frame_state.frame_obus = gf_list_new();
+		gf_list_add(state->frame_state.frame_obus, a);
+	}
+}
+
+GF_Err aom_av1_parse_obu_from_section5(GF_BitStream *bs, AV1State *state) {
```

Comment:
The name is a bit misleading as it parses all OBUs until if finds a temporal delimiter. So it parses a Temporal Unit. I'd suggest renaming it: `aom_av1_parse_tu_from_section5_obus`